## Analysis of the models performance

In [1]:
import pandas as pd

### result from the test grid

In [70]:
df = pd.read_csv('./results/05-29-2019-19:05:43.csv')

#### get the best model of each matrics

In [10]:
lst = ['auc-roc','accuracy_at_1', 'accuracy_at_2', 'accuracy_at_5',
       'accuracy_at_10', 'accuracy_at_20', 'accuracy_at_30', 'accuracy_at_50',
       'precision_at_1', 'precision_at_2', 'precision_at_5', 'precision_at_10',
       'precision_at_20', 'precision_at_30', 'precision_at_50', 'recall_at_1',
       'recall_at_2', 'recall_at_5', 'recall_at_10', 'recall_at_20',
       'recall_at_30', 'recall_at_50']
res_lst =[]
for item in lst:
    temp = df.sort_values(by = item,ascending = False).head(1)[['model_type','parameters','clf','train_start','train_end',item]]
    temp['metrics']  = item
    temp.rename(columns = {item:'score'},inplace =True)
    temp_1 = temp[['model_type','metrics', 'train_end','score']]
    res_lst.append(temp_1)
best_models = pd.concat(res_lst)

In [11]:
best_models

model_type          metrics   train_end     score
2          RF          auc-roc  2013-05-02  0.688447
16         RF    accuracy_at_1  2012-05-02  0.742931
16         RF    accuracy_at_2  2012-05-02  0.742263
16         RF    accuracy_at_5  2012-05-02  0.740352
16         RF   accuracy_at_10  2012-05-02  0.732646
16         RF   accuracy_at_20  2012-05-02  0.705522
2          RF   accuracy_at_30  2013-05-02  0.678402
9          RF   accuracy_at_50  2012-11-02  0.615976
2          RF   precision_at_1  2013-05-02  0.594104
9          RF   precision_at_2  2012-11-02  0.559908
9          RF   precision_at_5  2012-11-02  0.553917
9          RF  precision_at_10  2012-11-02  0.540092
9          RF  precision_at_20  2012-11-02  0.505413
10         AB  precision_at_30  2012-11-02  0.480498
9          RF  precision_at_50  2012-11-02  0.431862
2          RF      recall_at_1  2013-05-02  0.020840
2          RF      recall_at_2  2013-05-02  0.038498
16         RF      recall_at_5  2012-05-02  0.091993
16         RF     recall_at_10  2012-05-02  0.174303
2          RF     recall_at_20  2013-05-02  0.321031
2          RF     recall_at_30  2013-05-02  0.462059
19         BG     recall_at_50  2012-05-02  0.712919

#### As we can tell, RF overperforms than the others in almost every k percentage level.  From the table above, the best model of every matix is embedding model. The common models can't overperfom than the embedding models in this project. Since the results are based on the testing grid, the ranking may change when we use the small grid and large grid.

### Each matix's top 10 models

In [43]:
lst = ['auc-roc','accuracy_at_1', 'accuracy_at_2', 'accuracy_at_5',
       'accuracy_at_10', 'accuracy_at_20', 'accuracy_at_30', 'accuracy_at_50',
       'precision_at_1', 'precision_at_2', 'precision_at_5', 'precision_at_10',
       'precision_at_20', 'precision_at_30', 'precision_at_50', 'recall_at_1',
       'recall_at_2', 'recall_at_5', 'recall_at_10', 'recall_at_20',
       'recall_at_30', 'recall_at_50']
res_dic = {}
for item in lst:
    res_dic[item] = df.sort_values(by = item,ascending = False).head(10)[['model_type','parameters','clf','train_start','train_end',item]]

### top 10 models for auc-roc

In [45]:
res_dic['auc-roc']

model_type                                         parameters  \
2          RF  {'max_depth': 10, 'max_features': 'sqrt', 'min...   
9          RF  {'max_depth': 10, 'max_features': 'sqrt', 'min...   
10         AB        {'algorithm': 'SAMME', 'n_estimators': 100}   
16         RF  {'max_depth': 10, 'max_features': 'sqrt', 'min...   
6          GB  {'learning_rate': 0.1, 'max_depth': 10, 'n_est...   
3          AB        {'algorithm': 'SAMME', 'n_estimators': 100}   
17         AB        {'algorithm': 'SAMME', 'n_estimators': 100}   
1          LR                       {'C': 0.01, 'penalty': 'l1'}   
0          DT  {'criterion': 'gini', 'max_depth': 10, 'min_sa...   
7          DT  {'criterion': 'gini', 'max_depth': 10, 'min_sa...   

                                                  clf train_start   train_end  \
2   RandomForestClassifier(bootstrap=True, class_w...  2012-01-01  2013-05-02   
9   RandomForestClassifier(bootstrap=True, class_w...  2012-01-01  2012-11-02   
10  AdaBoostClassifier(algorithm='SAMME',\n       ...  2012-01-01  2012-11-02   
16  RandomForestClassifier(bootstrap=True, class_w...  2012-01-01  2012-05-02   
6   GradientBoostingClassifier(criterion='friedman...  2012-01-01  2013-05-02   
3   AdaBoostClassifier(algorithm='SAMME',\n       ...  2012-01-01  2013-05-02   
17  AdaBoostClassifier(algorithm='SAMME',\n       ...  2012-01-01  2012-05-02   
1   LogisticRegression(C=0.01, class_weight=None, ...  2012-01-01  2013-05-02   
0   DecisionTreeClassifier(class_weight=None, crit...  2012-01-01  2013-05-02   
7   DecisionTreeClassifier(class_weight=None, crit...  2012-01-01  2012-11-02   

     auc-roc  
2   0.688447  
9   0.682569  
10  0.676769  
16  0.670367  
6   0.667772  
3   0.666279  
17  0.657055  
1   0.649891  
0   0.649453  
7   0.640871

### top 10 models for precision_at_5

In [48]:
res_dic['precision_at_5']

model_type                                         parameters  \
9          RF  {'max_depth': 10, 'max_features': 'sqrt', 'min...   
10         AB        {'algorithm': 'SAMME', 'n_estimators': 100}   
8          LR                       {'C': 0.01, 'penalty': 'l1'}   
2          RF  {'max_depth': 10, 'max_features': 'sqrt', 'min...   
3          AB        {'algorithm': 'SAMME', 'n_estimators': 100}   
16         RF  {'max_depth': 10, 'max_features': 'sqrt', 'min...   
6          GB  {'learning_rate': 0.1, 'max_depth': 10, 'n_est...   
1          LR                       {'C': 0.01, 'penalty': 'l1'}   
4          ET  {'criterion': 'gini', 'max_depth': 10, 'max_fe...   
11         ET  {'criterion': 'gini', 'max_depth': 10, 'max_fe...   

                                                  clf train_start   train_end  \
9   RandomForestClassifier(bootstrap=True, class_w...  2012-01-01  2012-11-02   
10  AdaBoostClassifier(algorithm='SAMME',\n       ...  2012-01-01  2012-11-02   
8   LogisticRegression(C=0.01, class_weight=None, ...  2012-01-01  2012-11-02   
2   RandomForestClassifier(bootstrap=True, class_w...  2012-01-01  2013-05-02   
3   AdaBoostClassifier(algorithm='SAMME',\n       ...  2012-01-01  2013-05-02   
16  RandomForestClassifier(bootstrap=True, class_w...  2012-01-01  2012-05-02   
6   GradientBoostingClassifier(criterion='friedman...  2012-01-01  2013-05-02   
1   LogisticRegression(C=0.01, class_weight=None, ...  2012-01-01  2013-05-02   
4   ExtraTreesClassifier(bootstrap=False, class_we...  2012-01-01  2013-05-02   
11  ExtraTreesClassifier(bootstrap=False, class_we...  2012-01-01  2012-11-02   

    precision_at_5  
9         0.553917  
10        0.533641  
8         0.519816  
2         0.509058  
3         0.490036  
16        0.472694  
6         0.472373  
1         0.471920  
4         0.468750  
11        0.461751

### top 10 models for recall_at_5

In [49]:
res_dic['recall_at_5']

model_type                                         parameters  \
16         RF  {'max_depth': 10, 'max_features': 'sqrt', 'min...   
2          RF  {'max_depth': 10, 'max_features': 'sqrt', 'min...   
9          RF  {'max_depth': 10, 'max_features': 'sqrt', 'min...   
3          AB        {'algorithm': 'SAMME', 'n_estimators': 100}   
10         AB        {'algorithm': 'SAMME', 'n_estimators': 100}   
6          GB  {'learning_rate': 0.1, 'max_depth': 10, 'n_est...   
1          LR                       {'C': 0.01, 'penalty': 'l1'}   
4          ET  {'criterion': 'gini', 'max_depth': 10, 'max_fe...   
8          LR                       {'C': 0.01, 'penalty': 'l1'}   
20         GB  {'learning_rate': 0.1, 'max_depth': 10, 'n_est...   

                                                  clf train_start   train_end  \
16  RandomForestClassifier(bootstrap=True, class_w...  2012-01-01  2012-05-02   
2   RandomForestClassifier(bootstrap=True, class_w...  2012-01-01  2013-05-02   
9   RandomForestClassifier(bootstrap=True, class_w...  2012-01-01  2012-11-02   
3   AdaBoostClassifier(algorithm='SAMME',\n       ...  2012-01-01  2013-05-02   
10  AdaBoostClassifier(algorithm='SAMME',\n       ...  2012-01-01  2012-11-02   
6   GradientBoostingClassifier(criterion='friedman...  2012-01-01  2013-05-02   
1   LogisticRegression(C=0.01, class_weight=None, ...  2012-01-01  2013-05-02   
4   ExtraTreesClassifier(bootstrap=False, class_we...  2012-01-01  2013-05-02   
8   LogisticRegression(C=0.01, class_weight=None, ...  2012-01-01  2012-11-02   
20  GradientBoostingClassifier(criterion='friedman...  2012-01-01  2012-05-02   

    recall_at_5  
16     0.091993  
2      0.089405  
9      0.087648  
3      0.086064  
10     0.084439  
6      0.082962  
1      0.082883  
4      0.082326  
8      0.082252  
20     0.080657

#### Conclusion for results of the testing grid

- From the results with the testing grid, we can find out that the accuracy is not a good matrix, the top 10 models for accuracy are not as same as top 10 for recall and for precision.

- In term of using different training and testing dataset for the same model, the performance of the model doesn't change much.

### result from the small grid

In [51]:
df1 = pd.read_csv('./results/05-30-2019-02:31:45.csv')
df2 = pd.read_csv('./results/05-30-2019-03:00:58.csv')
df3 = pd.read_csv('./results/05-30-2019-03:15:56.csv')
res = [df1, df2, df3]
res_df = pd.concat(res)

In [53]:
lst = ['auc-roc','accuracy_at_1', 'accuracy_at_2', 'accuracy_at_5',
       'accuracy_at_10', 'accuracy_at_20', 'accuracy_at_30', 'accuracy_at_50',
       'precision_at_1', 'precision_at_2', 'precision_at_5', 'precision_at_10',
       'precision_at_20', 'precision_at_30', 'precision_at_50', 'recall_at_1',
       'recall_at_2', 'recall_at_5', 'recall_at_10', 'recall_at_20',
       'recall_at_30', 'recall_at_50']
res_lst =[]
for item in lst:
    temp = res_df.sort_values(by = item,ascending = False).head(1)[['model_type','parameters','clf','train_start','train_end',item]]
    temp['metrics']  = item
    temp.rename(columns = {item:'score'},inplace =True)
    temp_1 = temp[['model_type','metrics', 'train_end','score']]
    res_lst.append(temp_1)
best_models = pd.concat(res_lst)

#### Get the best models from the model pool for each matrix

In [54]:
best_models

model_type          metrics   train_end     score
43         AB          auc-roc  2013-05-02  0.684780
26         LR    accuracy_at_1  2012-05-02  0.743780
35         RF    accuracy_at_2  2012-05-02  0.742506
29         RF    accuracy_at_5  2012-05-02  0.739624
40         AB   accuracy_at_10  2012-05-02  0.730218
45         AB   accuracy_at_20  2012-05-02  0.705158
35         RF   accuracy_at_30  2013-05-02  0.677995
12         DT   accuracy_at_50  2012-11-02  0.761828
35         RF   precision_at_1  2012-11-02  0.672811
35         RF   precision_at_2  2012-11-02  0.645161
24         LR   precision_at_5  2012-11-02  0.592627
43         AB  precision_at_10  2012-11-02  0.548387
43         AB  precision_at_20  2012-11-02  0.506105
39         AB  precision_at_30  2012-11-02  0.481112
12         DT  precision_at_50  2012-11-02  0.577720
35         RF      recall_at_1  2012-11-02  0.021292
35         RF      recall_at_2  2012-11-02  0.040834
24         LR      recall_at_5  2012-11-02  0.093773
45         AB     recall_at_10  2013-05-02  0.176901
35         RF     recall_at_20  2013-05-02  0.326678
35         RF     recall_at_30  2013-05-02  0.461343
12         DT     recall_at_50  2012-05-02  1.000000

- Clearly, comparing to the best model from the testing grid, this time, RF doesn't domainate the ranking as before.
- In general, the embedding models are still better than the common ones.
- Since we care about the two matrics, one is auc-roc, the other is precision.We need to further investagate the top 10 of auc-roc, since this time, the best model for auc-roc is AD, however, the best model for precision_at_5 is LR.


In [55]:
lst = ['auc-roc','accuracy_at_1', 'accuracy_at_2', 'accuracy_at_5',
       'accuracy_at_10', 'accuracy_at_20', 'accuracy_at_30', 'accuracy_at_50',
       'precision_at_1', 'precision_at_2', 'precision_at_5', 'precision_at_10',
       'precision_at_20', 'precision_at_30', 'precision_at_50', 'recall_at_1',
       'recall_at_2', 'recall_at_5', 'recall_at_10', 'recall_at_20',
       'recall_at_30', 'recall_at_50']
res_dic = {}
for item in lst:
    res_dic[item] = res_df.sort_values(by = item,ascending = False).head(10)[['model_type','parameters','clf','train_start','train_end',item]]

### top 10 models for auc-roc

In [57]:
res_dic['auc-roc']

model_type                                         parameters  \
43         AB      {'algorithm': 'SAMME.R', 'n_estimators': 100}   
35         RF  {'max_depth': 50, 'max_features': 'sqrt', 'min...   
31         RF  {'max_depth': 5, 'max_features': 'sqrt', 'min_...   
44         AB     {'algorithm': 'SAMME.R', 'n_estimators': 1000}   
29         RF  {'max_depth': 5, 'max_features': 'sqrt', 'min_...   
45         AB    {'algorithm': 'SAMME.R', 'n_estimators': 10000}   
33         RF  {'max_depth': 50, 'max_features': 'sqrt', 'min...   
39         AB       {'algorithm': 'SAMME', 'n_estimators': 1000}   
38         AB        {'algorithm': 'SAMME', 'n_estimators': 100}   
29         RF  {'max_depth': 5, 'max_features': 'sqrt', 'min_...   

                                                  clf train_start   train_end  \
43  AdaBoostClassifier(algorithm='SAMME.R',\n     ...  2012-01-01  2013-05-02   
35  RandomForestClassifier(bootstrap=True, class_w...  2012-01-01  2013-05-02   
31  RandomForestClassifier(bootstrap=True, class_w...  2012-01-01  2013-05-02   
44  AdaBoostClassifier(algorithm='SAMME.R',\n     ...  2012-01-01  2013-05-02   
29  RandomForestClassifier(bootstrap=True, class_w...  2012-01-01  2013-05-02   
45  AdaBoostClassifier(algorithm='SAMME.R',\n     ...  2012-01-01  2013-05-02   
33  RandomForestClassifier(bootstrap=True, class_w...  2012-01-01  2013-05-02   
39  AdaBoostClassifier(algorithm='SAMME.R',\n     ...  2012-01-01  2012-11-02   
38  AdaBoostClassifier(algorithm='SAMME.R',\n     ...  2012-01-01  2012-11-02   
29  RandomForestClassifier(bootstrap=True, class_w...  2012-01-01  2012-11-02   

     auc-roc  
43  0.684780  
35  0.684754  
31  0.682265  
44  0.682042  
29  0.681948  
45  0.680413  
33  0.677933  
39  0.677682  
38  0.677189  
29  0.674747

### top 10 models for precision_at_5

In [59]:
res_dic['precision_at_5']

model_type                                         parameters  \
24         LR                          {'C': 1, 'penalty': 'l1'}   
25         LR                          {'C': 1, 'penalty': 'l2'}   
27         LR                         {'C': 10, 'penalty': 'l2'}   
26         LR                         {'C': 10, 'penalty': 'l1'}   
23         LR                        {'C': 0.1, 'penalty': 'l2'}   
43         AB      {'algorithm': 'SAMME.R', 'n_estimators': 100}   
22         LR                        {'C': 0.1, 'penalty': 'l1'}   
33         RF  {'max_depth': 50, 'max_features': 'sqrt', 'min...   
35         RF  {'max_depth': 50, 'max_features': 'sqrt', 'min...   
39         AB       {'algorithm': 'SAMME', 'n_estimators': 1000}   

                                                  clf train_start   train_end  \
24  LogisticRegression(C=10, class_weight=None, du...  2012-01-01  2012-11-02   
25  LogisticRegression(C=10, class_weight=None, du...  2012-01-01  2012-11-02   
27  LogisticRegression(C=10, class_weight=None, du...  2012-01-01  2012-11-02   
26  LogisticRegression(C=10, class_weight=None, du...  2012-01-01  2012-11-02   
23  LogisticRegression(C=10, class_weight=None, du...  2012-01-01  2012-11-02   
43  AdaBoostClassifier(algorithm='SAMME.R',\n     ...  2012-01-01  2012-11-02   
22  LogisticRegression(C=10, class_weight=None, du...  2012-01-01  2012-11-02   
33  RandomForestClassifier(bootstrap=True, class_w...  2012-01-01  2012-11-02   
35  RandomForestClassifier(bootstrap=True, class_w...  2012-01-01  2012-11-02   
39  AdaBoostClassifier(algorithm='SAMME.R',\n     ...  2012-01-01  2012-11-02   

    precision_at_5  
24        0.592627  
25        0.584332  
27        0.583410  
26        0.582488  
23        0.580645  
43        0.579724  
22        0.578802  
33        0.559447  
35        0.553917  
39        0.552995

### top 10 models for recall_at_5

In [60]:
res_dic['recall_at_5']

model_type                                         parameters  \
24         LR                          {'C': 1, 'penalty': 'l1'}   
25         LR                          {'C': 1, 'penalty': 'l2'}   
27         LR                         {'C': 10, 'penalty': 'l2'}   
26         LR                         {'C': 10, 'penalty': 'l1'}   
35         RF  {'max_depth': 50, 'max_features': 'sqrt', 'min...   
23         LR                        {'C': 0.1, 'penalty': 'l2'}   
43         AB      {'algorithm': 'SAMME.R', 'n_estimators': 100}   
22         LR                        {'C': 0.1, 'penalty': 'l1'}   
44         AB     {'algorithm': 'SAMME.R', 'n_estimators': 1000}   
45         AB    {'algorithm': 'SAMME.R', 'n_estimators': 10000}   

                                                  clf train_start   train_end  \
24  LogisticRegression(C=10, class_weight=None, du...  2012-01-01  2012-11-02   
25  LogisticRegression(C=10, class_weight=None, du...  2012-01-01  2012-11-02   
27  LogisticRegression(C=10, class_weight=None, du...  2012-01-01  2012-11-02   
26  LogisticRegression(C=10, class_weight=None, du...  2012-01-01  2012-11-02   
35  RandomForestClassifier(bootstrap=True, class_w...  2012-01-01  2013-05-02   
23  LogisticRegression(C=10, class_weight=None, du...  2012-01-01  2012-11-02   
43  AdaBoostClassifier(algorithm='SAMME.R',\n     ...  2012-01-01  2012-11-02   
22  LogisticRegression(C=10, class_weight=None, du...  2012-01-01  2012-11-02   
44  AdaBoostClassifier(algorithm='SAMME.R',\n     ...  2012-01-01  2013-05-02   
45  AdaBoostClassifier(algorithm='SAMME.R',\n     ...  2012-01-01  2013-05-02   

    recall_at_5  
24     0.093773  
25     0.092460  
27     0.092314  
26     0.092169  
35     0.092030  
23     0.091877  
43     0.091731  
22     0.091585  
44     0.091075  
45     0.090916

#### Get the auc-roc score for the best model of precision_at_5.

In [69]:
res_df[res_df['model_type'] == 'LR'].sort_values(by = 'precision_at_5',ascending = False).head(1)['auc-roc']

24    0.653305
Name: auc-roc, dtype: float64

#### Conclusion of the results of the small grid

- The best model for precision_at_5 is LR with parameters {'C': 1, 'penalty': 'l1'}. 
- Its auc-roc is 0.653305, which is not far away from the auc-roc score of the best model.
- We can use LR model for the project if the budget is enough for 5% of the whole project.
- For diffent K, the best model will be totally different, for example, at k = 20, RF would be the best choice.

In [21]:
from IPython.display import HTML

HTML('''<script>
code_show=true; 
function code_toggle() {
 if (code_show){
 $('div.input').hide();
 } else {
 $('div.input').show();
 }
 code_show = !code_show
} 
$( document ).ready(code_toggle);
</script>
<form action="javascript:code_toggle()"><input type="submit" value="Click here to toggle on/off the raw code."></form>''')